### PR TITLE
Slash Authorities for irrefutable misbehavior

### PR DESCRIPTION
**Pull request by rphmeier**
Authorities automatically generate and sign misbehavior report transactions for any misbehavior they've witnessed during the BFT agreement process.

Slashing right now just cuts the corresponding validator's bond down to zero and gives a portion of it to the reporter. Purging the offending authority will be more difficult.

### CODE REVIEW

**Comment by rphmeier** [comment]
(holding off on any more runtime changes for now since we'll probably merge in the governance model from #79)

File: polkadot/consensus/src/lib.rs [hunk 1]
```diff
@@ -565,6 +570,65 @@ impl<C: PolkadotApi, R: TableRouter> bft::Proposer for Proposer<C, R> {
...
+		};
+
+		for (target, misbehavior) in misbehavior {
+			let report = MisbehaviorReport {
+				parent_hash: self.parent_hash,
+				parent_number: self.parent_number,
+				target,
+				misbehavior: match misbehavior {
+					GenericMisbehavior::ProposeOutOfTurn(_, _, _) => continue,
+					GenericMisbehavior::DoublePropose(_, _, _) => continue,
```
Comment (rphmeier) [inline polkadot/consensus/src/lib.rs]: these are blocked on https://github.com/paritytech/polkadot/issues/85

**Comment by arkpar** [comment]
LGTM

File: polkadot/runtime/src/runtime/staking.rs [hunk 2]
```diff
@@ -21,19 +21,14 @@ use rstd::cell::RefCell;
 use runtime_io::print;
 use codec::KeyedVec;
 use runtime_support::{storage, StorageVec};
-use polkadot_primitives::{BlockNumber, AccountId};
-use runtime::{system, session, governance};
-
-/// The balance of an account.
-pub type Balance = u64;
```
Comment (gavofyork) [inline polkadot/runtime/src/runtime/staking.rs]: `Balance` and `Bonding` are types specific to staking and not polkadot.

**Comment by rphmeier** [comment]
@gavofyork restored the types. the thing I worry about is having duplicate type definitions (as we'll need one for transaction construction as well) and accidentally not changing one when we change the other.

**Pull request merged**